exp/clusterresourceset/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,380 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ClusterResourceSetReconciler reconciles a ClusterResourceSet object
+type ClusterResourceSetReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	config   *rest.Config
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=clusterresourceset.cluster.x-k8s.io,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=clusterresourceset.cluster.x-k8s.io,resources=clusterresourcesets/status,verbs=get;update;patch
+
+func (r *ClusterResourceSetReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+
+	// Fetch the ClusterResourceSet instance.
+	clusterResourceSet := &clusterresourcesetv1.ClusterResourceSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, clusterResourceSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(clusterResourceSet, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSet object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSet); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, clusterResourceSet)
+}
+
+// reconcile handles cluster reconciliation.
+func (r *ClusterResourceSetReconciler) reconcile(ctx context.Context, clusterResourceSet *clusterresourcesetv1.ClusterResourceSet) (ctrl.Result, error) {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	cls, err := r.getClustersFromClusterResourceSet(ctx, clusterResourceSet)
+	if err != nil {
+		logger.Error(err, "Failed fetching clusters that matches ClusterResourceSet labels", "ClusterResourceSet", clusterResourceSet.Name)
+		return ctrl.Result{}, err
+	}
+
+	for _, cl := range cls {
+		if err := r.ApplyClusterResourceSet(ctx, cl, clusterResourceSet); err != nil {
+			logger.Error(err, "Failed applying resources to cluster", "Cluster", cl.Name)
```

Comment:
If cluster resource failed applying , can we enqueue the request again so that it tries again. Otherwise, currently it waits for 5 mins to apply it again.

Reply:
Here, apply returns with error in a few scenarios: 1) if could not get remote cluster client, 2) if cannot patch status at the end, and 3) can not get ClusterResourceSetBinding object.

We can keep it as is to give it time to recover or requeue on error. Let's wait a bit to get feedback on this from others too.